browser/ai_chat/code_execution_tool.cc

```diff
@@ -0,0 +1,207 @@
+
+}  // namespace
+
+class CodeSandboxWebContentsObserver : public content::WebContentsObserver {
+ public:
+  using LoadCompleteCallback =
+      base::OnceCallback<void(content::RenderFrameHost*)>;
+
+  CodeSandboxWebContentsObserver(content::WebContents* web_contents,
+                                 LoadCompleteCallback callback)
+      : content::WebContentsObserver(web_contents),
+        load_complete_callback_(std::move(callback)) {}
+
+  ~CodeSandboxWebContentsObserver() override = default;
+
+  void DidFinishLoad(content::RenderFrameHost* render_frame_host,
+                     const GURL& validated_url) override {
+    // If the frame has a parent, we can assume this is the iframe
+    if (render_frame_host->GetParent() && load_complete_callback_) {
+      base::SingleThreadTaskRunner::GetCurrentDefault()->PostTask(
+          FROM_HERE, base::BindOnce(std::move(load_complete_callback_),
+                                    render_frame_host));
+    }
+  }
+
+ private:
+  LoadCompleteCallback load_complete_callback_;
+};
+
+CodeExecutionTool::CodeExecutionRequest::CodeExecutionRequest(
+    std::unique_ptr<content::WebContents> web_contents,
+    std::string wrapped_js,
+    UseToolCallback callback)
+    : web_contents(std::move(web_contents)),
+      wrapped_js(std::move(wrapped_js)),
+      callback(std::move(callback)) {}
+
+CodeExecutionTool::CodeExecutionRequest::~CodeExecutionRequest() = default;
+
+CodeExecutionTool::CodeExecutionTool(content::BrowserContext* browser_context)
+    : profile_(Profile::FromBrowserContext(browser_context)) {}
+
+CodeExecutionTool::~CodeExecutionTool() = default;
+
+std::string_view CodeExecutionTool::Name() const {
+  return mojom::kCodeExecutionToolName;
+}
+
+std::string_view CodeExecutionTool::Description() const {
+  return "Execute JavaScript code and return a human-readable formatted "
+         "string as a result. "
+         "Do not use console.log statements or similar statements. Always "
+         "return a string as a result."
+         "The code will be executed in a sandboxed environment.";
+}
+
+std::optional<base::Value::Dict> CodeExecutionTool::InputProperties() const {
+  return CreateInputProperties(
+      {{"script", StringProperty("The JavaScript code to execute")}});
+}
+
+std::optional<std::vector<std::string>> CodeExecutionTool::RequiredProperties()
+    const {
+  return std::vector<std::string>{"script"};
+}
+
+std::variant<bool, mojom::PermissionChallengePtr>
+CodeExecutionTool::RequiresUserInteractionBeforeHandling(
+    const mojom::ToolUseEvent& tool_use) const {
+  return false;
+}
+
+bool CodeExecutionTool::SupportsConversation(
+    bool is_temporary,
+    bool has_untrusted_content,
+    mojom::ConversationCapability conversation_capability) const {
+  // Support all conversation types for now
+  return true;
+}
+
+void CodeExecutionTool::SetExecutionTimeLimitForTesting(
+    base::TimeDelta time_limit) {
+  execution_time_limit_for_testing_ = time_limit;
```

Comment:
`execution_time_limit_for_testing_` should just be `execution_time_limit_` that you initialize in the constructor.

in the test you either call `SetExecutionTimeLimitForTesting` or initialize the timeout via constructor.

this part shouldn't exist in this case:
```
  auto time_limit =
      execution_time_limit_for_testing_.value_or(kExecutionTimeLimit);
```

Reply:
5debd6c26465f7fa8ac33641cf419996cae958ef